sips/cluster_consensus.md

```diff
@@ -0,0 +1,330 @@
+
+The `Cluster` will hold a `ClusterRunner` object for each slot it is running a beacon duty for.
+
+For other duty roles the old design will remain.
+
+#### Code
+
+```go
+// Cluster is a cluster of a unique set of operators that run the same validator set
+type Cluster interface {
+	// Initializes and starts the runner for the duties for the given slot
+  	Start(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
+	// ProcessMessage processes a message routed to this Cluster
+	ProcessMessage(msg *types.SSVMessage)
+}
+
+type Cluster struct {
+	Runners     map[spec.Slot]ClusterRunner
+    Shares      map[ValidatorPubkey]Share
+	Network           Network
+	Beacon            BeaconNode
+	OperatorID        OperatorID
+	QBFTParams        QBFTParams
+    
+}
+
+type QBFTParams interface {
+    Operators()        []OperatorID
+    Quorum()           uint64
+    PartialQuorum() 	  uint64
+}
+
+
+type Share struct {
+    // New field:
+	// HighestAttestingSlot holds the highest slot for which attester duty ran for a validator
+    HighestAttestingSlot spec.Slot
+}
+
+// ClusterRunner manages the duty cycle for a certain slot
+type ClusterRunner interface {
+	// Start the duty lifecycle for the given slot. Emits a message.
+    StartDuties(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
+	// Processes cosensus message
+    ProcessConsensus(consensusMessage *qbft.Message) error
+	// Processes a post-consensus message
+    ProcessPostConsensus(msg PartialSignatureMessages) 
+}
+
+type ClusterRunner struct {
+	Shares          map[ValidatorPubkey]Share
+    QBFTParams     QBFTParams
+	QBFTController *qbft.Controller
+	BeaconNetwork  *types.BeaconNetwork
+}
+
+// BeaconVote is the consensus data
+type BeaconVote struct {
+    BlockRoot         phase0.Root
+    Source            phase0.Checkpoint
+    Target            phase0.Checkpoint
+}
+
+func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
+    bits := bitfield.New(duty.CommitteeLength)
+    bits.Set(duty.ValidatorCommitteeIndex)
+
+    return Attestation{
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        Source: vote.Source,
+        Target: vote.Target,
+        CommiteeIndex: duty.CommitteeIndex,
+        AggregationBits: bits,
+    }
+}
+
+func ConstructSyncCommittee(vote BeaconVote, duty AttesterDuty) SyncCommitteeMessage{
+    return SyncCommitteeMessage
+    {
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        ValidatorIndex: duty.ValidatorIndex,
+    }
+}
+```
+#### PartialSignatureMessages
+
+The current structure that we have in code can be unchanged.
+
+```go
+type PartialSignatureMessages struct {
+	Type     PartialSigMsgType
+	Slot     phase0.Slot
+	Messages []*PartialSignatureMessage
+}
```

Comment:
So, indeed, the share's signatures will be merged into the same message, right? If so, what will be the maximum number of signatures per message?

Reply:
Max validators an operator can hold times 2 I believe